provisioning/device/src/ProvisioningDeviceClient.cs

```diff
@@ -63,52 +64,89 @@ public class ProvisioningDeviceClient
         /// <summary>
         /// Registers the current device using the Device Provisioning Service and assigns it to an IoT Hub.
         /// </summary>
+        /// <param name="timeout">The maximum amount of time to allow this operation to run for before timing out.</param>
+        /// <remarks>
+        /// Due to the AMQP library used by this library uses not accepting cancellation tokens, this overload and <see cref="RegisterAsync(ProvisioningRegistrationAdditionalData, TimeSpan)"/>
+        /// are the only overloads for this method that allow for a specified timeout to be respected in the middle of an AMQP operation such as opening
+        /// the AMQP connection. MQTT and HTTPS connections do not share that same limitation, though.
+        /// </remarks>
         /// <returns>The registration result.</returns>
-        public Task<DeviceRegistrationResult> RegisterAsync()
+        public Task<DeviceRegistrationResult> RegisterAsync(TimeSpan timeout)
         {
-            return RegisterAsync(CancellationToken.None);
+            Logging.RegisterAsync(this, _globalDeviceEndpoint, _idScope, _transport, _security);
+
+            return RegisterAsync(null, timeout);
         }
 
         /// <summary>
         /// Registers the current device using the Device Provisioning Service and assigns it to an IoT Hub.
         /// </summary>
-        /// <param name="data">The optional additional data.</param>
+        /// <param name="data">
+        /// The optional additional data that is passed through to the custom allocation policy webhook if 
+        /// a custom allocation policy webhook is setup for this enrollment.
+        /// </param>
+        /// <param name="timeout">The maximum amount of time to allow this operation to run for before timing out.</param>
+        /// <remarks>
+        /// Due to the AMQP library used by this library uses not accepting cancellation tokens, this overload and <see cref="RegisterAsync(TimeSpan)"/>
+        /// are the only overloads for this method that allow for a specified timeout to be respected in the middle of an AMQP operation such as opening
+        /// the AMQP connection. MQTT and HTTPS connections do not share that same limitation, though.
+        /// </remarks>
         /// <returns>The registration result.</returns>
-        public Task<DeviceRegistrationResult> RegisterAsync(ProvisioningRegistrationAdditionalData data)
+        public Task<DeviceRegistrationResult> RegisterAsync(ProvisioningRegistrationAdditionalData data, TimeSpan timeout)
```

Comment:
I was tempted to make the TimeSpan an optional parameter here like the cancellation token is, but we can't have two overloads with the same first parameter and different optional parameters

```csharp
public Task<DeviceRegistrationResult> RegisterAsync(ProvisioningRegistrationAdditionalData data, TimeSpan timeout = default);
public Task<DeviceRegistrationResult> RegisterAsync(ProvisioningRegistrationAdditionalData data, CancellationToken cancellationToken = default);
```

```csharp
public void myMethod()
{
    provisioningDeviceClient.RegisterAsync(data); // Which of the two above options would the compiler choose?
}
```

Reply:
Hmm

Could always add TimeSpan as another optional parameter to the one that has a CancellationToken. Would that be less weird?

Reply:
I could go either way on this suggestion. If we ever get cancellation token support for our AMQP library, then we can just deprecate all the APIs that take TimeSpans, but if we start mixing TimeSpans and Cancellation tokens then it gets messy.

Reply:
I'd like to keep the timespan and cancellation token APIs separate. Since both are ways to signal an API that it should stop the in-progress operation gracefully, I don't think we should mix them together.